### PR TITLE
Support function overloads

### LINKED ISSUE #30: support overloaded functions
Currently, the constructs like expectTypeOf.toBeCallableWith, expectTypeOf.parameters, expectTypeOf.parameter do not work with function overloads.  Basically they just assume the signature of the last overload and give error otherwise.

I believe this could be made work, doing something similar to what is described here:

https://stackoverflow.com/questions/59535995/parameters-generic-of-overloaded-function-doesnt-contain-all-options/59538756#59538756

Although, for expectTypeOf.parameter, I m not sure how to refer to a parameter at given index for specific overload.  But i believe "toBeCallableWith" and "parameters" could be made work.

Thanks.

### PR DESCRIPTION
**Pull request by aryaemami59**
## This PR:

  - [X] Adds support for function overloads through `toBeCallableWith()`
  - [x] Resolves #30 
  
  This is a rough draft, I will probably add support for overloads through `.parameter(number)` and `.parameters` as well.

### CODE REVIEW

**Comment by mmkal** [comment]
@aryaemami59, looking forward to this. Just in case there's anything breaking added in this, I would like hold off on v1 until this goes in, if you think it's close to being ready for review. But let's try as hard as possible to make sure it's non-breaking.

FYI I added you as a collaborator - next PR you open you should be able to just push a branch against this origin rather than your fork.

**Comment by aryaemami59** [comment]
@mmkal Thank you! I'm gonna make sure this is non-breaking, if it turns out it has to be a breaking change, then I might just make a new assertion instead.

**Comment by aryaemami59** [comment]
BTW I haven't forgotten about this, I'm just in the middle of moving. As soon as I'm done I will make sure to finish this.

**Comment by aryaemami59** [comment]
Not sure about this one so far, gonna need some feedback on this.

**Review by mmkal — COMMENTED** [review]
This looks good so far. I can't remember if we've already discussed whether this should also take over helpers like `.parameters` and `.returns`

e.g.

```ts
function f(n: number): 1
function f(n: number, s: string): 2
function f(...) {
  ...
}

expectTypeOf(f).parameters.toEqualTypeOf<[number] | [number, string]>()
expectTypeOf(f).returns.toEqualTypeOf<1 | 2>()
```

What do you think?

File: src/index.ts [hunk 1]
```diff
@@ -743,6 +776,166 @@ export type ExpectTypeOf<Actual, Options extends {positive: boolean}> = Options[
   ? PositiveExpectTypeOf<Actual>
   : NegativeExpectTypeOf<Actual>
 
+/**
+ * Any function with any arguments.
+ *
+ * @internal
+ */
+export type AnyFunction = (...args: any[]) => any
```
Comment (mmkal) [inline src/index.ts:784]: this doesn't seem to be used

File: src/index.ts [hunk 2]
```diff
@@ -743,6 +776,166 @@ export type ExpectTypeOf<Actual, Options extends {positive: boolean}> = Options[
...
+                    ? [(...args: A1) => R1]
+                    : any
+
+/**
+ * Extracts a tuple of parameter types from each overload
+ * of a given function type.
+ *
+ * @template FunctionType - The function type to extract parameter types from.
+ */
+export type OverloadedParameters<FunctionType> =
```
Comment (mmkal) [inline src/index.ts:928]: Can we add a "unit test" to this? i.e. something in `types.test.ts` or a new file that shows what this util actually does

**Review by mmkal — COMMENTED** [review]
Couple more comments

**Comment by aryaemami59** [comment]
> This looks good so far. I can't remember if we've already discussed whether this should also take over helpers like `.parameters` and `.returns`

Honestly, not too sure yet, it's up to you :)

**Comment by mmkal** [comment]
@aryaemami59 is there a reason the 10-overloads option needs to be the first ternary check rather than the last? I don't think this behaviour is great, even for an internal type:

image

[(playground, copy-pasting the `type Overloads = ...` definition you've got on this branch)](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=2&pc=1#code/MYewdgzgLgBAHgLhgeQG4FMBOAbEBDAEwgB4AKAShgF4A+GARjqpgG0Lq76BdAKB-TgAHEJlhQAnoPQoMOfEWIAxAK5hgUAJbgAKpPRMYKtZp16YAqOjBEYAbx4wYpAHSu8mAOYQkGsADMsGABBenIff0CAJXoHJ1dndy9wgMxggCYwmF8UmEi02Jc3T28siNSggGZM7KiKgvjEkprygBZqstyW+qKk0pyggFZ2nMiB7oTi5MCggDZhqJnxxqnygHZ51MjVpcm+6YAODdz9nd7m4IBOI8iL06aOkIAGa-pHngBfWIB+VljHR0KE16IUotFy9AANH9-oDlulQXQ8lD-jCGrtKgjchVkSi4j0SkE2hxOjiUbD0UNiaNSaj8UhZpjIjMaQC0cD1lTViy8UCCYcqftueTgVcqRchWyCa9Ga8abxHEgjOotGBdFJzHBLNYIHZoWTJStgqFDdE9bTeYaghkTflcea4ecMSa6nbWXS9q1rl1XTyHQ8hiaxj7hfd+nMTYtgwaPcF1ibtlH3Y7DiaTomLTGglcTbc7Z8UT8WGa3RmQVTIcXfeiMlS0tyS3CMVTsZWQ-SiWDIi161XgZTOwMe23gnMqczW9GghzO1yJ+6gvzO4K56XRZ3xWb5f9FaplaZ1RYrDZ7D6G7tHcaY6bT73Q9NrVfbafh46qs7K2ezg82ibvc-o46AZXkG-5Jg84ZXpGoEZo6cZXgm0F+v0KZXmmrr5rihYfre9KhOWPafgSNadnW2HDk2nYtje5Edoi3ZkZO-aIoODHzqOnbjtRk7Tois5cfOi6IsuPpbiiO7GCqarSIe2q6jehGWpe5zXvJL4PA+ylPvxMEPG+V4uqpAHfl62EKZmQHKSB2lIdMEHKVB1nng8cHKQhPoYXaWHyThRoygR9rVoypHeeRVTNv5ZmEoy9EhYxjIsbFbGMpxhnzjxuR8aeom4uJe6qmYMnHqZAVfv0SkdCpiU6f0GkVVpqXVdMenKQZVU2Z6v7FWZgHXFZDXtSO1wOaeHmul53k+WWnYVhNk3EYiwUTaFjJUUtk60SSXWTUxuQJWtSVjhFMDZXauUmPlB6akeOonrNPkXi8W3dep1z1W1Tn9M1FWtftjUdVef6zWp-QWRVfU3qNPqFuReHTRCc1BfDy3hZNG1dkjcVUgMJ2umdkkFVdsm3Xd90POVIwxCTz01a9T0lXe5RfSMP1A0Z-Q-gDdOQ6e0OTrDiKQgjtYY-OYWUSLpZoy0OM+nj+7SYTRVU-TimPcr1P3rT6uk591wsxN3M3rz878+CEuNvNuR1pNYuIhUMunnLF0K1qSvazr0zk1ElPu8Dms2nTKKG-Jxulqb0Tm4FtYOzeTtSRqrs6n75Sm8pPvK6Hjbh9wgfbjAYDoLIQA)

Makes it awkward to use for things like `expectTypeOf(myFunc).returns.toEqual<1 | 2>()`

**Comment by aryaemami59** [comment]
@mmkal I wish present me could remember why past me decided to go with that approach :) I'll look at it to see if I can change it.

**Comment by mmkal** [comment]
Oh right, of course, if you put the 1-overload variant TypeScript will just pick the last one semi-arbitrarily. But I now don't see what value we're getting from the intermediate ones. It seems we _might_ be stuck with a very long tuple as the output, but all functions will get a match from the first ternary. [Simplified demo](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAYgjFAvFAFAQwFxTgSiQPmwCgiBjAewDsBnYKAfQDMFkVKBXAG06jWt8ohe-APIA3CACdO5NABNqAJgA88fHmppgAS2qNtEfgG10WXAWwAaVJmx5EhOAF0SoSLEVIoAbyJR-NmY4Zn4BplCKwRFEAL5kVLQMjJ6sHNzCAkJ8UOJSMvJKqorqUJo6egbG4eYOVoER9oSKLkQQAB5g5JJ0btC50rIKKjCEyDBQ7cAQlAo+of4oAHTLaJIA5tRY2pSMUlAAgrhbO3sASnDzqMuLqxvHu5IHkfdnirGXAPxQJte3mwc1QjnaxLFbrf77SIWU7NS5YSgQCSSIhAA)

**Comment by aryaemami59** [comment]
@mmkal right, I was trying to see if there is a way to find out exactly how many overloads a function has so we don't run into that issue.

**Pull request closed without merging**